lang/ruby/lib/avro/io.rb

```diff
@@ -106,6 +106,38 @@ def read(len)
         @reader.read(len)
       end
 
+      def read_array_start
+        read_item_count
+      end
+
+      def read_array_next
+        read_item_count
+      end
+
+      def read_map_start
+        read_item_count
+      end
+
+      def read_map_next
+        read_item_count
+      end
+
+      def read_index
+        read_long
+      end
+
+      def read_enum
+        read_int
+      end
+
+      def read_item_count
+        block_count = read_long
+        if block_count < 0
+          block_count = -block_count
```

Comment:
If the block count is negative, then the size of the block in bytes is encoded afterwards. It looks like that was accidentally removed, which breaks reads because the size would be read in place of the next data value.